xc/device.ml

```diff
@@ -1134,56 +1148,76 @@ module PCI = struct
-                        |> String.split_on_char '\n'
-        in
-        List.iteri (fun i addr ->
-            if i < _proc_pci_num_resources then
-              Scanf.sscanf addr "0x%llx 0x%llx 0x%llx\n" (fun scan_start scan_end scan_flags ->
-                  if scan_start <> 0 then
-                    let size = scan_end - scan_start + 1 in
-                    let scan_start = if scan_flags <> 0 then scan_start else scan_start lsr 12 in
-                    let size = if size <> 0 then size else (_page_size + size - 1) lsr 12 in
-                    Xenctrl.domain_iomem_permission xc domid
-                      (Nativeint.of_int scan_start) (Nativeint.of_int size) true
-                )
-          )
-          addresses;
-        (sysfs_pci_dev ^ (Pci.string_of_address guest))
-        |> Unixext.string_of_file
-        |> Int64.of_string
-        |> Xenctrlext.physdev_map_pirq xc domid
-        |> fun x -> Xenctrl.domain_irq_permission xc domid x true
-      end
-    else
-      raise (Domain_not_running (host, domid))
+    if hvm then begin
+      if (Qemu.is_running ~xs domid) then
+        begin
+          let id = Printf.sprintf "pci-pt-%02x_%02x.%01x" host.bus host.dev host.fn in
+          let _qmp_result = qmp_send_cmd domid
+              (Qmp.Device_add {driver="xen-pci-passthrough";
+                               device=Qmp.Device.PCI({id;
+                                                      dev=guest.dev; fn=guest.fn;
+                                                      hostaddr=string_of_address host;
+                                                      permissive=false})}) in
+          ()
+        end
+      else
+        raise (Domain_not_running (host, domid));
+    end;
+    let addresses = (sysfs_pci_dev ^ (string_of_address host) ^ "/resource")
+                    |> Unixext.string_of_file
+                    |> String.split_on_char '\n'
+    in
+    List.iteri (fun i addr ->
+        if i < _proc_pci_num_resources then
+          Scanf.sscanf addr "0x%nx 0x%nx 0x%nx" (fun scan_start scan_end scan_flags ->
+              if scan_start <> 0n then
+                let scan_size = Nativeint.(sub scan_end scan_start |> succ) in
+                if Nativeint.(logand scan_flags _pci_bar_io > 0n) then begin
+                  Xenctrl.domain_ioport_permission xc domid (Nativeint.to_int scan_start)
+                    (Nativeint.to_int scan_size) true
+                end else begin
+                  let scan_start = Nativeint.(shift_right_logical scan_start 12) in
+                  let scan_size = Nativeint.(shift_right_logical (add _page_size scan_size |> pred) 12) in
+                  Xenctrl.domain_iomem_permission xc domid scan_start scan_size true
+                end
+            )
+      )
+      addresses;
+    let irq = (sysfs_pci_dev ^ (Pci.string_of_address host) ^ "/irq")
+              |> Unixext.string_of_file |> String.trim
+              |> int_of_string in
```

Comment:
Do we need to protect us against parsing failures here? On the other hand, we would not be able to recover so it is probably ok to let the exception escape.

Reply:
libxl would silently ignore parser failures https://github.com/mirage/xen/blob/master/tools/libxl/libxl_pci.c#L1076, but I wouldn't do that. I moved the parsing earlier in the function so that we would fail before we make any calls to Xen or Qemu.